src/main.cpp

```diff
@@ -17,13 +17,48 @@
 #include <cloysterhpc/services/shell.h>
 #include <cloysterhpc/verification.h>
 #include <cloysterhpc/view/newt.h>
+#include <cloysterhpc/functions.h>
+#include <cloysterhpc/dbus_client.h>
 #include <internal_use_only/config.hpp>
 #include <regex>
 
 #ifdef _CLOYSTER_I18N
 #include "include/i18n-cpp.hpp"
 #endif
 
+namespace {
+void initializeSingletons(auto&& cluster)
+{
+    using cloyster::models::Cluster;
+    cloyster::Singleton<Cluster>::init(
+        std::forward<decltype(cluster)>(cluster));
+
+    cloyster::Singleton<cloyster::services::BaseRunner>::init([](){
+        using cloyster::services::BaseRunner;
+        using cloyster::services::DryRunner;
+        using cloyster::services::Runner;
+
+        if (cloyster::dryRun) {
+            return cloyster::makeUniqueDerived<BaseRunner, DryRunner>();
+        }
+
+        return cloyster::makeUniqueDerived<BaseRunner, Runner>();
+    });
+
+    using cloyster::services::repos::RepoManager;
+    cloyster::Singleton<RepoManager>::init([]() {
+        auto clusterPtr = cloyster::Singleton<Cluster>::get();
+        const auto& osinfo = clusterPtr->getHeadnode().getOS();
+        return std::make_unique<RepoManager>(osinfo);
+    });
+
+    cloyster::Singleton<MessageBus>::init([]() {
+        return cloyster::makeUniqueDerived<MessageBus, DBusClient>(
+            "org.freedesktop.systemd1", "/org/freedesktop/systemd1");
+    });
+}
+}; // anonymous namespace 
+
```

Comment:
I'm using Singleton pattern explicitly now, for a class `C`, `Singleton<C>::init()` initializes the singleton, this is done once, in the main (which calls this function). Then we can do `Singleton<C>::get()` to get an instance.

`Singleton<C>::init()` has two overloads, in the first case it accepts the unique pointer that will be moved into the singleton, and in the second case it agrees with a factory function that returns the unique pointer to be moved to the singleton.

And this `cloyster::makeUniqueDerived<BaseRunner, DryRunner>()` is just a utility function to create unique pointers of a base type, it is literally the same as `static_cast<std::unique_ptr<BaseRunner>>(std::make_unique<DryRunner>())` which is too long